samples/nrf9160/nrf_cloud_mqtt_multi_service/src/connection.c

```diff
@@ -0,0 +1,648 @@
+/*
+ * Copyright (c) 2022 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-Nordic-5-Clause
+ */
+
+#include <zephyr.h>
+#include <stdio.h>
+#include <modem/lte_lc.h>
+#include <net/socket.h>
+#include <net/nrf_cloud.h>
+#include <date_time.h>
+#include <logging/log.h>
+#include <cJSON.h>
+
+#include "connection.h"
+#include "fota_support.h"
+
+LOG_MODULE_REGISTER(connection, CONFIG_MQTT_MULTI_SERVICE_LOG_LEVEL);
+
+/* Flow control event identifiers */
+
+/* LTE either is connected or isn't */
+#define LTE_CONNECTED		(1 << 1)
+
+/* Cloud notifies on connection loss, and this is responded to by connection manager */
+/* Do not confuse these with the similarly-named features of the generic cloud API */
+#define CLOUD_CONNECTED			(1 << 1)
+#define CLOUD_READY			(1 << 2)
+#define CLOUD_DISCONNECTED		(1 << 3)
+
+/* Time either is or is not known. This is only fired once, and is never cleared */
+#define DATE_TIME_KNOWN			(1 << 1)
+
+/* Flow control event objects for waiting for key events */
+static K_EVENT_DEFINE(lte_connection_events);
+static K_EVENT_DEFINE(cloud_connection_events);
+static K_EVENT_DEFINE(datetime_connection_events);
+
+/* Message Queue for enqueing outgoing messages during offline periods */
+K_MSGQ_DEFINE(device_message_queue, sizeof(char *), CONFIG_MAX_OUTGOING_MESSAGES, sizeof(char *));
+
+/**
+ * @brief Notify that LTE connection has been established
+ *
+ */
+static void notify_lte_connected(void)
+{
+	k_event_post(&lte_connection_events, LTE_CONNECTED);
+}
+
+/**
+ * @brief Reset the LTE connection event flag
+ */
+static void clear_lte_connected(void)
+{
+	k_event_set(&lte_connection_events, 0);
+}
+
+bool await_lte_connection(k_timeout_t timeout)
+{
+	LOG_DBG("Awaiting LTE Connection");
+	return k_event_wait_all(&lte_connection_events, LTE_CONNECTED, false, timeout);
```

Comment:
It's a little odd to me that `k_event_wait_all` returns a `uint32_t`, but you're implicitly casting it to bool.  It technically works because it returns 0 if no events fire, and 0 is usually the same as false.

Reply:
My understanding is that the implicit cast to bool should also convert any nonzero to "true". I can definitely see how this might harm readability.

But you're probably right, comparing with `0` would improve readability